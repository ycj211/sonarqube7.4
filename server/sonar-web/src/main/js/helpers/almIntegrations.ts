/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { CurrentUser, AlmOrganization } from '../app/types';

export function hasAdvancedALMIntegration(user: CurrentUser) {
  return (
    isLoggedIn(user) && (isBitbucket(user.externalProvider) || isGithub(user.externalProvider))
  );
}

export function isBitbucket(almKey?: string) {
  return almKey && almKey.startsWith('bitbucket');
}

export function isGithub(almKey?: string) {
  return almKey === 'github';
}

export function isVSTS(almKey?: string) {
  return almKey === 'microsoft';
}

export function isPersonal(organization?: AlmOrganization) {
  return Boolean(organization && organization.personal);
}

export function sanitizeAlmId(almKey?: string) {
  if (isBitbucket(almKey)) {
    return 'bitbucket';
  }
  return almKey;
}
