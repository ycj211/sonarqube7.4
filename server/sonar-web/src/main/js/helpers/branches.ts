/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sortBy } from 'lodash';
import {
  BranchLike,
  Branch,
  BranchType,
  ShortLivingBranch,
  LongLivingBranch,
  PullRequest,
  MainBranch,
  BranchParameters
} from '../app/types';

export function isBranch(branchLike?: BranchLike): branchLike is Branch {
  return branchLike !== undefined && (branchLike as Branch).isMain !== undefined;
}

export function isShortLivingBranch(branchLike?: BranchLike): branchLike is ShortLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as ShortLivingBranch).type === BranchType.SHORT
  );
}

export function isLongLivingBranch(branchLike?: BranchLike): branchLike is LongLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as LongLivingBranch).type === BranchType.LONG
  );
}

export function isMainBranch(branchLike?: BranchLike): branchLike is MainBranch {
  return isBranch(branchLike) && branchLike.isMain;
}

export function isPullRequest(branchLike?: BranchLike): branchLike is PullRequest {
  return branchLike !== undefined && (branchLike as PullRequest).key !== undefined;
}

export function getPullRequestDisplayName(pullRequest: PullRequest) {
  return `${pullRequest.key} â€“ ${pullRequest.title}`;
}

export function getBranchLikeDisplayName(branchLike: BranchLike) {
  return isPullRequest(branchLike) ? getPullRequestDisplayName(branchLike) : branchLike.name;
}

export function getBranchLikeKey(branchLike: BranchLike) {
  return isPullRequest(branchLike) ? `pull-request-${branchLike.key}` : `branch-${branchLike.name}`;
}

export function getBranchQualityGateColor(status: string) {
  let indicatorColor = 'gray';
  if (status === 'ERROR') {
    indicatorColor = 'red';
  } else if (status === 'WARN') {
    indicatorColor = 'orange';
  } else if (status === 'OK') {
    indicatorColor = 'green';
  }
  return indicatorColor;
}

export function isSameBranchLike(a: BranchLike | undefined, b: BranchLike | undefined) {
  // main branches are always equal
  if (isMainBranch(a) && isMainBranch(b)) {
    return true;
  }

  // short- and long-living branches are compared by type and name
  if (
    (isLongLivingBranch(a) && isLongLivingBranch(b)) ||
    (isShortLivingBranch(a) && isShortLivingBranch(b))
  ) {
    return a.type === b.type && a.name === b.name;
  }

  // pull requests are compared by id
  if (isPullRequest(a) && isPullRequest(b)) {
    return a.key === b.key;
  }

  // finally if both parameters are `undefined`, consider them equal
  return a === b;
}

export function sortBranchesAsTree(branchLikes: BranchLike[]) {
  const result: BranchLike[] = [];

  const mainBranch = branchLikes.find(isMainBranch);
  const longLivingBranches = branchLikes.filter(isLongLivingBranch);
  const shortLivingBranches = branchLikes.filter(isShortLivingBranch);
  const pullRequests = branchLikes.filter(isPullRequest);

  // main branch is always first
  if (mainBranch) {
    result.push(
      mainBranch,
      ...getPullRequests(mainBranch.name),
      ...getNestedShortLivingBranches(mainBranch.name)
    );
  }

  // then all long-living branches
  sortBy(longLivingBranches, 'name').forEach(longLivingBranch => {
    result.push(
      longLivingBranch,
      ...getPullRequests(longLivingBranch.name),
      ...getNestedShortLivingBranches(longLivingBranch.name)
    );
  });

  // finally all orhpan pull requests and branches
  result.push(
    ...sortBy(pullRequests.filter(pr => pr.isOrphan), pullRequest => pullRequest.key),
    ...sortBy(shortLivingBranches.filter(branch => branch.isOrphan), branch => branch.name)
  );

  return result;

  /** Get all short-living branches (possibly nested) which should be merged to a given branch */
  function getNestedShortLivingBranches(mergeBranch: string) {
    const found: ShortLivingBranch[] = shortLivingBranches.filter(
      branch => branch.mergeBranch === mergeBranch
    );

    let i = 0;
    while (i < found.length) {
      const current = found[i];
      found.push(...shortLivingBranches.filter(branch => branch.mergeBranch === current.name));
      i++;
    }

    return sortBy(found, branch => branch.name);
  }

  function getPullRequests(base: string) {
    return sortBy(pullRequests.filter(pr => pr.base === base), pullRequest => pullRequest.key);
  }
}

export function getBranchLikeQuery(branchLike?: BranchLike): BranchParameters {
  if (isShortLivingBranch(branchLike) || isLongLivingBranch(branchLike)) {
    return { branch: branchLike.name };
  } else if (isPullRequest(branchLike)) {
    return { pullRequest: branchLike.key };
  } else {
    return {};
  }
}

// Create branch object from branch name or pull request key
export function fillBranchLike(
  branch?: string,
  pullRequest?: string
): ShortLivingBranch | PullRequest | undefined {
  if (branch) {
    return {
      isMain: false,
      mergeBranch: '',
      name: branch,
      type: BranchType.SHORT
    } as ShortLivingBranch;
  } else if (pullRequest) {
    return { base: '', branch: '', key: pullRequest, title: '' } as PullRequest;
  }
  return undefined;
}
