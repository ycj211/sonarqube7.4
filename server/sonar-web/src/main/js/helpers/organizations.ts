/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { Organization, OrganizationSubscription, CurrentUser } from '../app/types';

export function isPaidOrganization(organization: Organization | undefined): boolean {
  return Boolean(organization && organization.subscription === OrganizationSubscription.Paid);
}

export function hasPrivateAccess(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return (
    !isPaidOrganization(organization) ||
    isCurrentUserMemberOf(currentUser, organization, userOrganizations)
  );
}

export function isCurrentUserMemberOf(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return Boolean(
    organization &&
      isLoggedIn(currentUser) &&
      ((organization.actions && organization.actions.admin) ||
        userOrganizations.some(org => org.key === organization.key))
  );
}
