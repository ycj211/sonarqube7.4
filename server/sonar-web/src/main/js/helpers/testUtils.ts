/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme';
import { InjectedRouter } from 'react-router';
import { IntlProvider } from 'react-intl';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', { ...mockEvent, ...event });
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.body.clientWidth/clientHeight` are getters by default, so we need to redefine them
  // pass `configurable: true` to allow to redefine the properties multiple times
  if (width) {
    Object.defineProperty(document.body, 'clientWidth', { configurable: true, value: width });
  }
  if (height) {
    Object.defineProperty(document.body, 'clientHeight', { configurable: true, value: height });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();
export function shallowWithIntl<C extends React.Component>(
  node: React.ReactElement<any>,
  options: ShallowRendererProps = {}
) {
  return shallow<C>(node, { ...options, context: { intl, ...options.context } });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}

export function mockRouter(overrides: { push?: Function; replace?: Function } = {}) {
  return {
    createHref: jest.fn(),
    createPath: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    isActive: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setRouteLeaveHook: jest.fn(),
    ...overrides
  } as InjectedRouter;
}
