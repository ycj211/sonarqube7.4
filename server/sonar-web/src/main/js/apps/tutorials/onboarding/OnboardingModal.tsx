/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import handleRequiredAuthentication from '../../../app/utils/handleRequiredAuthentication';
import Modal from '../../../components/controls/Modal';
import OnboardingPrivateIcon from '../../../components/icons-components/OnboardingPrivateIcon';
import OnboardingProjectIcon from '../../../components/icons-components/OnboardingProjectIcon';
import OnboardingTeamIcon from '../../../components/icons-components/OnboardingTeamIcon';
import { Button, ResetButtonLink } from '../../../components/ui/buttons';
import { translate } from '../../../helpers/l10n';
import { CurrentUser } from '../../../app/types';
import { getCurrentUser, Store } from '../../../store/rootReducer';
import { isLoggedIn } from '../../../helpers/users';
import '../styles.css';

interface OwnProps {
  onClose: () => void;
  onOpenOrganizationOnboarding: () => void;
  onOpenProjectOnboarding: () => void;
  onOpenTeamOnboarding: () => void;
}

interface StateProps {
  currentUser: CurrentUser;
}

type Props = OwnProps & StateProps;

export class OnboardingModal extends React.PureComponent<Props> {
  componentDidMount() {
    if (!isLoggedIn(this.props.currentUser)) {
      handleRequiredAuthentication();
    }
  }

  handleOpenProjectOnboarding = () => {
    this.props.onOpenProjectOnboarding();
  };

  render() {
    if (!isLoggedIn(this.props.currentUser)) {
      return null;
    }

    const header = translate('onboarding.header');
    return (
      <Modal
        contentLabel={header}
        medium={true}
        onRequestClose={this.props.onClose}
        shouldCloseOnOverlayClick={false}>
        <div className="modal-simple-head text-center">
          <h1>{translate('onboarding.header')}</h1>
          <p className="spacer-top">{translate('onboarding.header.description')}</p>
        </div>
        <div className="modal-simple-body text-center onboarding-choices">
          <Button className="onboarding-choice" onClick={this.handleOpenProjectOnboarding}>
            <OnboardingProjectIcon className="big-spacer-bottom" />
            <h6 className="onboarding-choice-name">
              {translate('onboarding.analyze_public_code')}
            </h6>
            <p className="note">{translate('onboarding.analyze_public_code.note')}</p>
          </Button>
          <Button className="onboarding-choice" onClick={this.props.onOpenOrganizationOnboarding}>
            <OnboardingPrivateIcon className="big-spacer-bottom" />
            <h6 className="onboarding-choice-name">
              {translate('onboarding.analyze_private_code')}
            </h6>
            <p className="note">{translate('onboarding.analyze_private_code.note')}</p>
          </Button>
          <Button className="onboarding-choice" onClick={this.props.onOpenTeamOnboarding}>
            <OnboardingTeamIcon className="big-spacer-bottom" />
            <h6 className="onboarding-choice-name">
              {translate('onboarding.contribute_existing_project')}
            </h6>
            <p className="note">{translate('onboarding.contribute_existing_project.note')}</p>
          </Button>
        </div>
        <div className="modal-simple-footer text-center">
          <ResetButtonLink className="spacer-bottom" onClick={this.props.onClose}>
            {translate('not_now')}
          </ResetButtonLink>
          <p className="note">{translate('onboarding.footer')}</p>
        </div>
      </Modal>
    );
  }
}

const mapStateToProps = (state: Store): StateProps => ({ currentUser: getCurrentUser(state) });

export default connect(mapStateToProps)(OnboardingModal);
