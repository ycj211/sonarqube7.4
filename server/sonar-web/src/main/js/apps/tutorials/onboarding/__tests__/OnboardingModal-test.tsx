/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { OnboardingModal } from '../OnboardingModal';
import { click } from '../../../../helpers/testUtils';

it('renders correctly', () => {
  expect(
    shallow(
      <OnboardingModal
        currentUser={{ isLoggedIn: true }}
        onClose={jest.fn()}
        onOpenOrganizationOnboarding={jest.fn()}
        onOpenProjectOnboarding={jest.fn()}
        onOpenTeamOnboarding={jest.fn()}
      />
    )
  ).toMatchSnapshot();
});

it('should correctly open the different tutorials', () => {
  const onClose = jest.fn();
  const onOpenOrganizationOnboarding = jest.fn();
  const onOpenProjectOnboarding = jest.fn();
  const onOpenTeamOnboarding = jest.fn();
  const push = jest.fn();
  const wrapper = shallow(
    <OnboardingModal
      currentUser={{ isLoggedIn: true }}
      onClose={onClose}
      onOpenOrganizationOnboarding={onOpenOrganizationOnboarding}
      onOpenProjectOnboarding={onOpenProjectOnboarding}
      onOpenTeamOnboarding={onOpenTeamOnboarding}
    />,
    { context: { router: { push } } }
  );

  click(wrapper.find('ResetButtonLink'));
  expect(onClose).toHaveBeenCalled();

  wrapper.find('Button').forEach(button => click(button));
  expect(onOpenOrganizationOnboarding).toHaveBeenCalled();
  expect(onOpenProjectOnboarding).toHaveBeenCalled();
  expect(onOpenTeamOnboarding).toHaveBeenCalled();
});
