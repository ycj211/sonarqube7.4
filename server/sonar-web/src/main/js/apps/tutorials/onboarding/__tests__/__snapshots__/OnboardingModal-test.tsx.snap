// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  contentLabel="onboarding.header"
  medium={true}
  onRequestClose={[MockFunction]}
  shouldCloseOnOverlayClick={false}
>
  <div
    className="modal-simple-head text-center"
  >
    <h1>
      onboarding.header
    </h1>
    <p
      className="spacer-top"
    >
      onboarding.header.description
    </p>
  </div>
  <div
    className="modal-simple-body text-center onboarding-choices"
  >
    <Button
      className="onboarding-choice"
      onClick={[Function]}
    >
      <OnboardingProjectIcon
        className="big-spacer-bottom"
      />
      <h6
        className="onboarding-choice-name"
      >
        onboarding.analyze_public_code
      </h6>
      <p
        className="note"
      >
        onboarding.analyze_public_code.note
      </p>
    </Button>
    <Button
      className="onboarding-choice"
      onClick={[MockFunction]}
    >
      <OnboardingPrivateIcon
        className="big-spacer-bottom"
      />
      <h6
        className="onboarding-choice-name"
      >
        onboarding.analyze_private_code
      </h6>
      <p
        className="note"
      >
        onboarding.analyze_private_code.note
      </p>
    </Button>
    <Button
      className="onboarding-choice"
      onClick={[MockFunction]}
    >
      <OnboardingTeamIcon
        className="big-spacer-bottom"
      />
      <h6
        className="onboarding-choice-name"
      >
        onboarding.contribute_existing_project
      </h6>
      <p
        className="note"
      >
        onboarding.contribute_existing_project.note
      </p>
    </Button>
  </div>
  <div
    className="modal-simple-footer text-center"
  >
    <ResetButtonLink
      className="spacer-bottom"
      onClick={[MockFunction]}
    >
      not_now
    </ResetButtonLink>
    <p
      className="note"
    >
      onboarding.footer
    </p>
  </div>
</Modal>
`;
